### ISSUE TITLE
Get Data back to the editor

### ISSUE THREAD

**Issue by sixtyseven-multimedia** [aut0poietic/wp-admin-modal-example#5]
Thank you for this fantastic example. I managed to incorporate the iframe modal into a plugin without any hassle.

Let's asume I have a form with radio buttons. Could you provide some help how to get the value of this radio (or other form field(s)) back to the editor?

```
$("#btn-ok").on('click', function( e ){
     e.preventDefault();
     alert('OK clicked!');
     //functionality here
     parent.my_dialog_close_handler( e );
 });
```

Greetings from Germany!

**Comment by sixtyseven-multimedia** [comment]
That's what I came up with so far:

```
$("#btn-ok").on('click', function( e ){
     e.preventDefault();

     $.ajax({
        type: "post",
        url: myplugin_modal_vars.ajaxurl, // comes from wp_localize_script via admin_url('admin-ajax.php')
        data: {
            action: 'myplugin_modal_submit_data', // comes from wp_localize_script
            submitted: $( '#myplugin_modal_form' ).serialize() // all form fields as a string
        }
    }).done(function (response) {
        parent.myplugin_dialog_close_handler( e );
    }).fail(function () {
        // Whoops; show an error.
        alert(myplugin_modal_vars.ajax_fail);
        parent.myplugin_dialog_close_handler( e );
    });

 });
```

That way I can give the form data to a class function via an ajax hook:

```
add_action( 'wp_ajax_myplugin_modal_submit_data', array($this, 'modal_frame_submitted') );
```

Here I can epxplode the submitted data, so that I have later on the values of each form field in a variable. I now only have one step left: Adding these values as a string to the WordPress tinymce Editor, preferred at the current cursor position. Any help with that is gladly appreciated.

P.S.: I could of course go the easiest way and add a tinymce button via the normal API and a (probably dynamically via php created) plugin call. But since I try to code a foundation for a later on easy extendable plugin, your iframe way would be the best method. Not to mention that I could extend it's content in a snap.

**Comment by aut0poietic** [comment]
Hey André, thanks for the shout out from Germany.  Oklahoma says 'Howdy' ;)

Working with TinyMCE is pretty far beyond what I was trying to accomplish with this example, but:

The route you're taking using $.ajax is not going to work if you're trying to update tinyMCE. TinyMCE's state is held on the front-end in JavaScript and HTML, while submitting the data using $.ajax will put the values in the PHP on the back-end. 

You'll need to work entirely in JavaScript, grabbing an instance of the parent frame then replace the current selection in TinyMCE. 

``` javascript
// This is psudo-code, but should mostly work

jQuery(function($){
    $('.my-button').on('click', function() {
        // do what you need to do with your form data
        // build a shortcode, create some html, whatever
        var myHTMLFragment = getMyContent(); 

        if( window.parent.tinyMCE ){
            window.parent.tinyMCE.activeEditor.selection. setContent( myHTMLFragment );
        }
    }
});
```

More Info:

http://www.tinymce.com/wiki.php/API3:class.tinymce.Editor
http://www.tinymce.com/wiki.php/API3:class.tinymce.dom.Selection 

If you're trying to integrate with TinyMCE, you may be much better off creating a TinyMCE Plugin and launching the modal from that plugin.  It may also save you some time if you used the Backbone method, as you wouldn't have to navigate the parent frame ( which I've found problematic in the past ).

This codex article should get you started. and there's no better examples of TinyMCE plugins than those within WordPress itself, the `hr` plugin is very simple; `media` is more advanced, but deals with shortcodes and content replacement in visual and text mode.

https://codex.wordpress.org/TinyMCE_Custom_Buttons
https://github.com/WordPress/WordPress/tree/master/wp-includes/js/tinymce/plugins

**Comment by sixtyseven-multimedia** [comment]
Howdy ;-)
I managed to get the ajax way working, but it was somewhat over dimensioned for my goal. Your answer, howewer, pushed me into the right direction. What I tried to achieve was a way to insert text/html to the editor or, if text mode is used, to the appropriate textarea, so a tinyMCE button simply wasn't enough. I decided to add a button next to the media button. The javascript for the modal content now is in fact a php script with wp-load included to get benefit of the hooks and the language. There I included a little jQuery function 'send_to_editor', which does the trick:

```
function send_to_content(sendstring, e){
var textfield = jQuery('#content',parent.document);
if(textfield.css('display') == 'block'){
    // texteditor is used
    var old_content = textfield.val();
    var new_content = old_content + sendstring;
    textfield.val(new_content);
} else {
    // wysiwyg is used
    var editor = parent.tinyMCE.get('content');
    editor.insertContent(sendstring);
}
// close modal
parent.dialog_close_handler( e );
}
```

By using parent.tinyMCE.get('content') and not parent.tinyMCE.activeEditor, the given string is inserted, regardless of the focus status of the editor. If is focussed, it will be inserted at the caret position, if not, it will simply be inserted at the very top.

There are a few glitches, though: I have to find a way to set the editor's cursor behind the END of the inserted content, now it is inside the tags before the closing one, when inserting for example  &lt;div id="mydiv"&gt;My Content&lt;/div&gt;
Second, I'd like to insert the content at the very bottom, if not focussed. We'll see about this later.

Regarding the textarea: It would be great, if I could insert the content at the caret position as well, but the field loses it's focus, when opening the modal iframe. So I have to put it after the actual content.

Lastly: Why am I doing all this? I'd like to have a dialog, where addons to my plugin could hook into later on. And when doing it your way (I still prefer the iframe variant, howewer), I could place whatever I want into this modal, since the modal content html and the action js is generated dynamically within WP.

Thanks again for the Idea and the helping hand.
